File: DSA/src/Questions/BackTracking/NQueens2.java
package Questions.BackTracking;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Set;
import java.util.HashSet;

/*
https://leetcode.com/problems/n-queens-ii/
The n-queens puzzle is the problem of placing n queens on an n x n chessboard such that no two queens attack each other.

Given an integer n, return the number of distinct solutions to the n-queens puzzle.

 

Example 1:


Input: n = 4
Output: 2
Explanation: There are two distinct solutions to the 4-queens puzzle as shown.
Example 2:

Input: n = 1
Output: 1
 

Constraints:

1 <= n <= 9
*/
public class NQueens2 {
    public int totalNQueens(int n) {
        // Array to keep track of the position of queens in each row
        int[] queens = new int[n];
        int[] count = new int[1];
        // Call the backtracking function starting from the first row
        backtrack(queens, 0, n, count);
        return count[0];
    }

    private void backtrack(int[] queens, int row, int n, int[] count) {
        // If all rows are filled, we found a valid solution
        if (row == n) {
            count[0]++;
            return;
        }
        // Try placing a queen in each column of the current row
        for (int col = 0; col < n; col++) {
            if (isSafe(queens, row, col)) {
                // Place the queen in the current position
                queens[row] = col;
                // Move to the next row
                backtrack(queens, row + 1, n, count);
                // Backtrack: remove the queen from the current position
                queens[row] = 0;
            }
        }
    }

    private boolean isSafe(int[] queens, int row, int col) {
        // Check if the current position is safe to place a queen
        for (int i = 0; i < row; i++) {
            int placedQueenCol = queens[i];
            // Check if the current column or diagonals are under attack
            if (placedQueenCol == col ||
                    placedQueenCol - i == col - row ||
                    placedQueenCol + i == col + row) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        NQueens2 nQueens2 = new NQueens2();
        System.out.println(nQueens2.totalNQueens(4));
    }

}

File: DSA/Python/Questions/BackTracking/NQueens2.py
"""
# Question: N-Queens II
# Link: https://leetcode.com/problems/n-queens-ii/

# Count total number of distinct N-Queens solutions

# Time Complexity: O(N!)
# Space Complexity: O(N)

# Algorithm:
# 1. Use backtracking to try queen placements
# 2. Track queen positions in each row
# 3. Validate diagonal and column conflicts
# 4. Count valid solutions

# Key Components:
# - total_n_queens(): Main solution counter
# - backtrack(): Recursive helper for queen placement
# - is_safe(): Position validation checker
"""


class NQueens2:
    def total_n_queens(self, n: int) -> int:
        queens = [0] * n
        count = [0]

        def is_safe(queens: list, row: int, col: int) -> bool:
            for i in range(row):
                placed_queen_col = queens[i]
                if (
                    placed_queen_col == col
                    or placed_queen_col - i == col - row
                    or placed_queen_col + i == col + row
                ):
                    return False
            return True

        def backtrack(queens: list, row: int, n: int, count: list):
            if row == n:
                count[0] += 1
                return

            for col in range(n):
                if is_safe(queens, row, col):
                    queens[row] = col
                    backtrack(queens, row + 1, n, count)
                    queens[row] = 0

        backtrack(queens, 0, n, count)
        return count[0]


def main():
    solution = NQueens2()
    print(solution.total_n_queens(4))  # Expected output: 2


if __name__ == "__main__":
    main()
